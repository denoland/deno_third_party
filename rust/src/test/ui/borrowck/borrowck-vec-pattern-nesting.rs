#![feature(box_patterns)]
#![feature(box_syntax)]
#![feature(slice_patterns)]

fn a() {
    let mut vec = [box 1, box 2, box 3];
    match vec {
        [box ref _a, _, _] => {
        //~^ borrow of `vec[..]` occurs here
            vec[0] = box 4; //~ ERROR cannot assign
            //~^ assignment to borrowed `vec[..]` occurs here
            _a.use_ref();
        }
    }
}

fn b() {
    let mut vec = vec![box 1, box 2, box 3];
    let vec: &mut [Box<isize>] = &mut vec;
    match vec {
        &mut [ref _b..] => {
        //~^ borrow of `vec[..]` occurs here
            vec[0] = box 4; //~ ERROR cannot assign
            //~^ assignment to borrowed `vec[..]` occurs here
            _b.use_ref();
        }
    }
}

fn c() {
    let mut vec = vec![box 1, box 2, box 3];
    let vec: &mut [Box<isize>] = &mut vec;
    match vec {
        &mut [_a, //~ ERROR cannot move out
            //~| cannot move out
            //~| to prevent move
            ..
        ] => {
            // Note: `_a` is *moved* here, but `b` is borrowing,
            // hence illegal.
            //
            // See comment in middle/borrowck/gather_loans/mod.rs
            // in the case covering these sorts of vectors.
        }
        _ => {}
    }
    let a = vec[0]; //~ ERROR cannot move out
    //~| cannot move out of here
}

fn d() {
    let mut vec = vec![box 1, box 2, box 3];
    let vec: &mut [Box<isize>] = &mut vec;
    match vec {
        &mut [ //~ ERROR cannot move out
        //~^ cannot move out
         _b] => {}
        _ => {}
    }
    let a = vec[0]; //~ ERROR cannot move out
    //~| cannot move out of here
}

fn e() {
    let mut vec = vec![box 1, box 2, box 3];
    let vec: &mut [Box<isize>] = &mut vec;
    match vec {
        &mut [_a, _b, _c] => {}  //~ ERROR cannot move out
        //~| cannot move out
        _ => {}
    }
    let a = vec[0]; //~ ERROR cannot move out
    //~| cannot move out of here
}

fn main() {}

trait Fake { fn use_mut(&mut self) { } fn use_ref(&self) { }  }
impl<T> Fake for T { }
