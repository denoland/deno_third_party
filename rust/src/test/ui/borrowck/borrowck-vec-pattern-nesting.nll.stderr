error[E0506]: cannot assign to `vec[_]` because it is borrowed
  --> $DIR/borrowck-vec-pattern-nesting.rs:10:13
   |
LL |         [box ref _a, _, _] => {
   |              ------ borrow of `vec[_]` occurs here
LL |         //~^ borrow of `vec[..]` occurs here
LL |             vec[0] = box 4; //~ ERROR cannot assign
   |             ^^^^^^ assignment to borrowed `vec[_]` occurs here
LL |             //~^ assignment to borrowed `vec[..]` occurs here
LL |             _a.use_ref();
   |             -- borrow later used here

error[E0506]: cannot assign to `vec[_]` because it is borrowed
  --> $DIR/borrowck-vec-pattern-nesting.rs:23:13
   |
LL |         &mut [ref _b..] => {
   |               ------ borrow of `vec[_]` occurs here
LL |         //~^ borrow of `vec[..]` occurs here
LL |             vec[0] = box 4; //~ ERROR cannot assign
   |             ^^^^^^ assignment to borrowed `vec[_]` occurs here
LL |             //~^ assignment to borrowed `vec[..]` occurs here
LL |             _b.use_ref();
   |             -- borrow later used here

error[E0508]: cannot move out of type `[std::boxed::Box<isize>]`, a non-copy slice
  --> $DIR/borrowck-vec-pattern-nesting.rs:33:11
   |
LL |     match vec {
   |           ^^^ cannot move out of here
LL |         &mut [_a, //~ ERROR cannot move out
   |               -- data moved here
   |
note: move occurs because `_a` has type `std::boxed::Box<isize>`, which does not implement the `Copy` trait
  --> $DIR/borrowck-vec-pattern-nesting.rs:34:15
   |
LL |         &mut [_a, //~ ERROR cannot move out
   |               ^^
help: consider removing the `&mut`
   |
LL |         [_a, //~ ERROR cannot move out
LL |             //~| cannot move out
LL |             //~| to prevent move
LL |             ..
LL |         ] => {
   |

error[E0508]: cannot move out of type `[std::boxed::Box<isize>]`, a non-copy slice
  --> $DIR/borrowck-vec-pattern-nesting.rs:47:13
   |
LL |     let a = vec[0]; //~ ERROR cannot move out
   |             ^^^^^^
   |             |
   |             cannot move out of here
   |             help: consider borrowing here: `&vec[0]`

error[E0508]: cannot move out of type `[std::boxed::Box<isize>]`, a non-copy slice
  --> $DIR/borrowck-vec-pattern-nesting.rs:54:11
   |
LL |     match vec {
   |           ^^^ cannot move out of here
...
LL |          _b] => {}
   |          -- data moved here
   |
note: move occurs because `_b` has type `std::boxed::Box<isize>`, which does not implement the `Copy` trait
  --> $DIR/borrowck-vec-pattern-nesting.rs:57:10
   |
LL |          _b] => {}
   |          ^^
help: consider removing the `&mut`
   |
LL |         [ //~ ERROR cannot move out
LL |         //~^ cannot move out
LL |          _b] => {}
   |

error[E0508]: cannot move out of type `[std::boxed::Box<isize>]`, a non-copy slice
  --> $DIR/borrowck-vec-pattern-nesting.rs:60:13
   |
LL |     let a = vec[0]; //~ ERROR cannot move out
   |             ^^^^^^
   |             |
   |             cannot move out of here
   |             help: consider borrowing here: `&vec[0]`

error[E0508]: cannot move out of type `[std::boxed::Box<isize>]`, a non-copy slice
  --> $DIR/borrowck-vec-pattern-nesting.rs:67:11
   |
LL |     match vec {
   |           ^^^ cannot move out of here
LL |         &mut [_a, _b, _c] => {}  //~ ERROR cannot move out
   |         -----------------
   |         |     |   |   |
   |         |     |   |   ...and here
   |         |     |   ...and here
   |         |     data moved here
   |         help: consider removing the `&mut`: `[_a, _b, _c]`
   |
note: move occurs because these variables have types that don't implement the `Copy` trait
  --> $DIR/borrowck-vec-pattern-nesting.rs:68:15
   |
LL |         &mut [_a, _b, _c] => {}  //~ ERROR cannot move out
   |               ^^  ^^  ^^

error[E0508]: cannot move out of type `[std::boxed::Box<isize>]`, a non-copy slice
  --> $DIR/borrowck-vec-pattern-nesting.rs:72:13
   |
LL |     let a = vec[0]; //~ ERROR cannot move out
   |             ^^^^^^
   |             |
   |             cannot move out of here
   |             help: consider borrowing here: `&vec[0]`

error: aborting due to 8 previous errors

Some errors occurred: E0506, E0508.
For more information about an error, try `rustc --explain E0506`.
